plat/common/arm/lcpu.c

```diff
@@ -42,6 +45,20 @@
 
 #define CPU_ID_MASK 0xff00ffffffUL
 
+/*
+ *       STACK_SIZE               STACK_SIZE
+ * <---------------------><--------------------->
+ * |============================================|
+ * |                     |                      |
+ * |       trap stack    |        IRQ stack     |
+ * |                     |                      |
+ * |=============================================
```

Comment:
Another perhaps simpler approach would be to define a single exception stack that uses `SP_EL0`. Then we can do `msr SPSel, #0` and solve the problem of whether the exception is nested or not. What do you think?

Reply:
I like the idea.

> and solve the problem of whether the exception is nested or not
I am not sure I understand how this would solve this, but what it definetely help us with is getting rid of the need for a scratch register. Good idea! Will try it out.

Reply:
Oh, man, unfortuantely, there are drawbacks. It does still work but it has its drawbacks. I am attaching here a diff for reference in case we ever have to come back to this (tested on Redis)
```diff
diff --git a/plat/kvm/arm/exceptions.S b/plat/kvm/arm/exceptions.S
index b02be2a9..e9c92032 100644
--- a/plat/kvm/arm/exceptions.S
+++ b/plat/kvm/arm/exceptions.S
@@ -88,10 +88,10 @@
 	 *                       ^
 	 *                     SP_EL0
 	 */
-	mrs     x0, sp_el0
+	mov     x0, sp
 
-	/* Make it so that SP contains SP_EL0 and x0 contains old SP */
-	EXCHANGE_SP_WITH_X0
+	msr	spsel, #0
+	mrs	x0, sp_el1
 
 .if \is_irq != 0
 	/* If we are an IRQ, use the IRQ stack instead. */
@@ -162,7 +162,7 @@
 	stp	x22, x23, [sp, #16 * 16]
 .endm
 
-.macro RESTORE_REGS
+.macro RESTORE_REGS, is_irq
 	/* Mask IRQ to make sure restore would not be interrupted by IRQ */
 	msr daifset, #2
 
@@ -181,7 +181,7 @@
 	ldp x24, x25, [sp, #16 * 12]
 	ldp x22, x23, [sp, #16 * 11]
 	ldp x20, x21, [sp, #16 * 10]
-	/* Skip x18, x19 */
+	ldp x18, x19, [sp, #16 * 9]
 	ldp x16, x17, [sp, #16 * 8]
 	ldp x14, x15, [sp, #16 * 7]
 	ldp x12, x13, [sp, #16 * 6]
@@ -192,13 +192,13 @@
 	ldp x2, x3, [sp, #16 * 1]
 	ldp x0, x1, [sp, #16 * 0]
 
-	/* Restore stack pointer for exception from EL1 */
-	ldr x18, [sp, #__SP_OFFSET]
-	mov x19, sp
-	mov sp, x18
+.if \is_irq != 0
+	/* If we are an IRQ, use the IRQ stack instead. */
+	add     sp, sp, #CPU_EXCEPT_STACK_SIZE
+.endif
 
-	/* Restore x18, x19 */
-	ldp x18, x19, [x19, #16 * 9]
+	/* Restore stack pointer for exception from EL1 */
+	msr	spsel, #0x1
 
 	eret
 .endm
@@ -214,7 +214,7 @@ el1_sync:
 	mov x0, sp
 	mrs x1, far_el1
 	bl trap_el1_sync
-	RESTORE_REGS
+	RESTORE_REGS 0
 
 .align 6
 el1_irq:
@@ -222,7 +222,7 @@ el1_irq:
 	msr daifclr, #(8 | 4 | 1)
 	mov x0, sp
 	bl trap_el1_irq
-	RESTORE_REGS
+	RESTORE_REGS 1
 
 /* Bad Abort numbers */
 #define BAD_SYNC  0
 ```
 
 But there are mainly two problems:
 - Once we write to `SPSel`, we also modify `SPSR_EL1` and the one that ends up being saved on the frame is not really the original one... although, yes, you can patch it afterwards knowing that this modification happened, but it is not nice looking at all
 - When we reach the point of return we must restore the stack to its previous value before restoring `SPSel` to `1` on normal execution, but in the case of a nested situation (e.g. trap happening during IRQ handling) we would need to check on return whether we really should restore SPSel or not, which ends up being not so nice looking either :cry: 
 
 All in all, this ends up resulting in having to do a tiny bit more instructions, however the biggest problem is design wise as we end up doing some ugly explicit patching and checking.

Reply:
Agreed, adding checks on the exit sequence defeats the purpose of that change, as the idea was to simplify the design with `SPSel`. Thanks a lot for looking into it 🙏🏼